trellis-http/src/main/java/org/trellisldp/http/impl/PutHandler.java

```diff
@@ -239,8 +237,24 @@ public ResponseBuilder setResource(final Resource res) {
                         .forEachOrdered(dataset::add);
                 }
             });
+            // TODO is this the best we can do? what concurrency errors are lurking?
+            Future<Boolean> success = resourceService.get(internalId).isPresent()
```

Comment:
Actually, this check here is overkill. A better formulation would be:

```java
Future<Boolean> success = nonNull(res)
                  ? resourceService.replace(internalId, ldpType, dataset.asDataset())
                  : resourceService.create(internalId, ldpType, dataset.asDataset());
```

Basically, this very same check is already performed at a higher level in `LdpResource`, and depending on the result it either calls `::setResource` with the retrieved `Resource` or it calls the no-argument `::setResource` (which is equivalent to calling `setResource(null)`).

If we want to be more aggressive about avoiding conflicting writes, we could require an `If-Match` header for all PUT-for-update requests. (LDP describes this at the level of `should` -- Trellis supports `If-Match` but doesn't require it for updates).

Reply:
That's good to hear (that the check can be so much less expensive). I'm not a big fan right now of adding restriction to LDP-- I'd rather leave collision avoidance to the backend for now unless we find out that everyone does it exactly the same way. (Which could very well happen...) Keep HTTP as stateless and unconversational as possible!